Setup Code for Periphials

Siri2K opened the pull request:
Setup I2C code to connect to BMS, PDS and BPL. Each periphial file runs the listed commands in Commands.h. For  now, none of functions have outputs, without testing.

Siri2K committed:
Setup Code for Periphials

Setup I2C code to connect to BMS, PDS and BPL. Each periphial file
runs the listed commands in Commands.h. For  now, none of functions
have outputs, without testing.

rima1881 commented:
it is good but i have some small changes. can you change the bpl to backplane since in all of our docs it is mentioned backplane and it is easier to identify it.

rima1881 commented:
the other thing is i2c. i am not so sure if it is a good idea to store slave address in i2c file. if some how we don't change it we will send commands to wrong devices. i think it is better to put the address in the argument or somehow include it in the enum that we are sending

rima1881 commented:
i am not sure if it is possible in c but we can make an struct that includes both address and enum command so we can just pass them at the same time

Siri2K commented:
> it is good but i have some small changes. can you change the bpl to backplane since in all of our docs it is mentioned backplane and it is easier to identify it.

Will do

Siri2K commented:
> the other thing is i2c. i am not so sure if it is a good idea to store slave address in i2c file. if some how we don't change it we will send commands to wrong devices. i think it is better to put the address in the argument or somehow include it in the enum that we are sending

> i am not sure if it is possible in c but we can make an struct that includes both address and enum command so we can just pass them at the same time


Actually, I think I have a suggestion to fix both of these. In Python, we can use JSON files to setup a dictionary to that stores hard-coded values, then read them in a Python. I'm think we can do something similar with C with a JSON, XML or another file type and then read the addresses from there. 

If we pull it off, all of the addresses (device and commands) can be soft-coded into the SAMV71 and then if we do change the addresses, we can just upload a new version of the database file without touching the code.

I'll still do it as you asked, but if you think this is a good idea, I can take of it as a new issue

Siri2K commented:
> the other thing is i2c. i am not so sure if it is a good idea to store slave address in i2c file. if some how we don't change it we will send commands to wrong devices. i think it is better to put the address in the argument or somehow include it in the enum that we are sending

The original Address are stored in the bms.h, pds.h, bpl.h. connectTo__ basically calls the I2CConnect function, we can probably just call `connectTo__` on every command to first connect to device, then execute the command?

Siri2K committed:
Changed BPL to Backplane

rima1881 commented:
> > the other thing is i2c. i am not so sure if it is a good idea to store slave address in i2c file. if some how we don't change it we will send commands to wrong devices. i think it is better to put the address in the argument or somehow include it in the enum that we are sending
> 
> The original Address are stored in the bms.h, pds.h, bpl.h. connectTo__ basically calls the I2CConnect function, we can probably just call `connectTo__` on every command to first connect to device, then execute the command?

yeah but the thing is if we are not careful with that reference we might cause it to send command to wrong device. for example if we do multi-threading and we run the command at the same time one of the address will be replaced so sending the address in each function is much more safe.

Siri2K commented:
Ok fair enough, I'll fix up the cod later today

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Amir ***@***.***>
Sent: Thursday, August 29, 2024 10:36:53 AM
To: scsd-cdh/Hades ***@***.***>
Cc: Sriram Kanagalingham ***@***.***>; Author ***@***.***>
Subject: Re: [scsd-cdh/Hades] Setup Code for Periphials (PR #20)


the other thing is i2c. i am not so sure if it is a good idea to store slave address in i2c file. if some how we don't change it we will send commands to wrong devices. i think it is better to put the address in the argument or somehow include it in the enum that we are sending

The original Address are stored in the bms.h, pds.h, bpl.h. connectTo__ basically calls the I2CConnect function, we can probably just call connectTo__ on every command to first connect to device, then execute the command?

yeah but the thing is if we are not careful with that reference we might cause it to send command to wrong device. for example if we do multi-threading and we run the command at the same time one of the address will be replaced so sending the address in each function is much more safe.

—
Reply to this email directly, view it on GitHub<https://github.com/scsd-cdh/Hades/pull/20#issuecomment-2317902796>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AQBYTCZDJEGUAJH23YFUI6DZT4WYLAVCNFSM6AAAAABNB4Z5NSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMJXHEYDENZZGY>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

rima1881 commented:
> > the other thing is i2c. i am not so sure if it is a good idea to store slave address in i2c file. if some how we don't change it we will send commands to wrong devices. i think it is better to put the address in the argument or somehow include it in the enum that we are sending
> 
> > i am not sure if it is possible in c but we can make an struct that includes both address and enum command so we can just pass them at the same time
> 
> Actually, I think I have a suggestion to fix both of these. In Python, we can use JSON files to setup a dictionary to that stores hard-coded values, then read them in a Python. I'm think we can do something similar with C with a JSON, XML or another file type and then read the addresses from there.
> 
> If we pull it off, all of the addresses (device and commands) can be soft-coded into the SAMV71 and then if we do change the addresses, we can just upload a new version of the database file without touching the code.
> 
> I'll still do it as you asked, but if you think this is a good idea, I can take of it as a new issue

I am not sure about dictionary but it's not a bad idea. storing it doesn't matter whether it is in json or something else. the problem with dictionary is that it needs to stay on ram all the time which is a little bit annoying or we need a lot of mram calls which is annoying as well

Siri2K commented:
True, plus we don't have any way to access the RAM indirectly.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Amir ***@***.***>
Sent: Thursday, August 29, 2024 10:39:32 AM
To: scsd-cdh/Hades ***@***.***>
Cc: Sriram Kanagalingham ***@***.***>; Author ***@***.***>
Subject: Re: [scsd-cdh/Hades] Setup Code for Periphials (PR #20)


the other thing is i2c. i am not so sure if it is a good idea to store slave address in i2c file. if some how we don't change it we will send commands to wrong devices. i think it is better to put the address in the argument or somehow include it in the enum that we are sending

i am not sure if it is possible in c but we can make an struct that includes both address and enum command so we can just pass them at the same time

Actually, I think I have a suggestion to fix both of these. In Python, we can use JSON files to setup a dictionary to that stores hard-coded values, then read them in a Python. I'm think we can do something similar with C with a JSON, XML or another file type and then read the addresses from there.

If we pull it off, all of the addresses (device and commands) can be soft-coded into the SAMV71 and then if we do change the addresses, we can just upload a new version of the database file without touching the code.

I'll still do it as you asked, but if you think this is a good idea, I can take of it as a new issue

I am not sure about dictionary but it's not a bad idea. storing it doesn't matter whether it is in json or something else. the problem with dictionary is that it needs to stay on ram all the time which is a little bit annoying or we need a lot of mram calls which is annoying as well

—
Reply to this email directly, view it on GitHub<https://github.com/scsd-cdh/Hades/pull/20#issuecomment-2317911815>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AQBYTC5CBMWQXSD6TIWSLGLZT4XCJAVCNFSM6AAAAABNB4Z5NSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMJXHEYTCOBRGU>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

rima1881 commented:
> True, plus we don't have any way to access the RAM indirectly. Get Outlook for Android<https://aka.ms/AAb9ysg>
> […](#)
> ________________________________ From: Amir ***@***.***> Sent: Thursday, August 29, 2024 10:39:32 AM To: scsd-cdh/Hades ***@***.***> Cc: Sriram Kanagalingham ***@***.***>; Author ***@***.***> Subject: Re: [scsd-cdh/Hades] Setup Code for Periphials (PR #20) the other thing is i2c. i am not so sure if it is a good idea to store slave address in i2c file. if some how we don't change it we will send commands to wrong devices. i think it is better to put the address in the argument or somehow include it in the enum that we are sending i am not sure if it is possible in c but we can make an struct that includes both address and enum command so we can just pass them at the same time Actually, I think I have a suggestion to fix both of these. In Python, we can use JSON files to setup a dictionary to that stores hard-coded values, then read them in a Python. I'm think we can do something similar with C with a JSON, XML or another file type and then read the addresses from there. If we pull it off, all of the addresses (device and commands) can be soft-coded into the SAMV71 and then if we do change the addresses, we can just upload a new version of the database file without touching the code. I'll still do it as you asked, but if you think this is a good idea, I can take of it as a new issue I am not sure about dictionary but it's not a bad idea. storing it doesn't matter whether it is in json or something else. the problem with dictionary is that it needs to stay on ram all the time which is a little bit annoying or we need a lot of mram calls which is annoying as well — Reply to this email directly, view it on GitHub<[#20 (comment)](https://github.com/scsd-cdh/Hades/pull/20#issuecomment-2317911815)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AQBYTC5CBMWQXSD6TIWSLGLZT4XCJAVCNFSM6AAAAABNB4Z5NSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMJXHEYTCOBRGU>. You are receiving this because you authored the thread.Message ID: ***@***.***>



> True, plus we don't have any way to access the RAM indirectly. Get Outlook for Android<https://aka.ms/AAb9ysg>
> […](#)
> ________________________________ From: Amir ***@***.***> Sent: Thursday, August 29, 2024 10:39:32 AM To: scsd-cdh/Hades ***@***.***> Cc: Sriram Kanagalingham ***@***.***>; Author ***@***.***> Subject: Re: [scsd-cdh/Hades] Setup Code for Periphials (PR #20) the other thing is i2c. i am not so sure if it is a good idea to store slave address in i2c file. if some how we don't change it we will send commands to wrong devices. i think it is better to put the address in the argument or somehow include it in the enum that we are sending i am not sure if it is possible in c but we can make an struct that includes both address and enum command so we can just pass them at the same time Actually, I think I have a suggestion to fix both of these. In Python, we can use JSON files to setup a dictionary to that stores hard-coded values, then read them in a Python. I'm think we can do something similar with C with a JSON, XML or another file type and then read the addresses from there. If we pull it off, all of the addresses (device and commands) can be soft-coded into the SAMV71 and then if we do change the addresses, we can just upload a new version of the database file without touching the code. I'll still do it as you asked, but if you think this is a good idea, I can take of it as a new issue I am not sure about dictionary but it's not a bad idea. storing it doesn't matter whether it is in json or something else. the problem with dictionary is that it needs to stay on ram all the time which is a little bit annoying or we need a lot of mram calls which is annoying as well — Reply to this email directly, view it on GitHub<[#20 (comment)](https://github.com/scsd-cdh/Hades/pull/20#issuecomment-2317911815)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AQBYTC5CBMWQXSD6TIWSLGLZT4XCJAVCNFSM6AAAAABNB4Z5NSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMJXHEYTCOBRGU>. You are receiving this because you authored the thread.Message ID: ***@***.***>

we can make it global but i would rather not

rima1881 commented:
i think what we sure do is to include the address in each function of devices or we can make i function like get address

rima1881 commented:
and pass it around when ever we need it. and ones done the address is discarded

rima1881 commented:
but give me a day to see other options

Siri2K commented:
Alright

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Amir ***@***.***>
Sent: Thursday, August 29, 2024 10:55:39 AM
To: scsd-cdh/Hades ***@***.***>
Cc: Sriram Kanagalingham ***@***.***>; Author ***@***.***>
Subject: Re: [scsd-cdh/Hades] Setup Code for Periphials (PR #20)


but give me a day to see other options

—
Reply to this email directly, view it on GitHub<https://github.com/scsd-cdh/Hades/pull/20#issuecomment-2317973306>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AQBYTC6AODE6E43OU4ZL42DZT4Y6XAVCNFSM6AAAAABNB4Z5NSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMJXHE3TGMZQGY>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

Siri2K commented:
Ok, I guess in terms of the mods, I'll do this today and we can fix the rest tmr:
- I'll figure out a way to put the instructions in enum and Address together in the way you originally suggested
- Drivers has a way to connect and disconnect between I2C, I can use that to validate and make sure the right address is receiving the commands

Siri2K committed:
Updated Periphials

Added some protection codes to check if any I2C function fail.
Changed Enum to Structs for Commands. Created Structs for PDS, BMS,
and BPL. Each Struct contains device address, io_descriptor for I2C,
and pointers to command respective structs. Added code to acquire
slave address

Commit:
Final Touches on the Periphial Code preTest
